rosbag2/src/rosbag2/player.cpp

```diff
@@ -0,0 +1,134 @@
+// Copyright 2018, Bosch Software Innovations GmbH.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "player.hpp"
+
+#include <chrono>
+#include <map>
+#include <memory>
+#include <queue>
+#include <string>
+#include <vector>
+#include <utility>
+
+#include "rclcpp/rclcpp.hpp"
+#include "rcl/graph.h"
+#include "rcutils/time.h"
+
+#include "rosbag2_storage/serialized_bag_message.hpp"
+#include "rosbag2_storage/storage_interfaces/read_only_interface.hpp"
+#include "rosbag2_storage/storage_factory.hpp"
+#include "rosbag2/logging.hpp"
+#include "rosbag2_node.hpp"
+#include "replayable_message.hpp"
+#include "typesupport_helpers.hpp"
+
+namespace rosbag2
+{
+
+Player::Player(std::shared_ptr<rosbag2_storage::storage_interfaces::ReadOnlyInterface> storage)
+: storage_(storage), node_(std::make_shared<Rosbag2Node>("rosbag2_node"))
+{}
+
+bool is_pending(const std::future<void> & future)
+{
+  return !(future.valid() && future.wait_for(std::chrono::seconds(0)) == std::future_status::ready);
+}
+
+void Player::play(Rosbag2PlayOptions options)
+{
+  prepare_publishers();
+
+  auto storage_loading_future = std::async(std::launch::async,
+      [this, options]() {load_storage_content(options);});
+
+  wait_for_filled_queue(options, storage_loading_future);
+
+  play_messages_from_queue(std::move(storage_loading_future));
+}
+
+void Player::wait_for_filled_queue(
+  const Rosbag2PlayOptions & options, const std::future<void> & storage_loading_future) const
+{
+  while (
+    message_queue_.size_approx() < options.queue_buffer_length_ &&
+    is_pending(storage_loading_future))
+  {
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
```

Comment:
I could see that this parameter should be exposed or configurable. Could we pack this into the `Rosbag2PlayOptions` struct?

Reply:
Definitely possible. I am not yet sure how useful this parameter will be. I suggest to discuss this here (#34).